hazelcast/reactor.py

```diff
@@ -83,13 +200,53 @@ def _check_timers(self):
                     # timers in the heap.
                     return
 
-    def add_timer_absolute(self, timeout, callback):
-        timer = Timer(timeout, callback)
-        self._new_timers.append((timer.end, timer))
-        return timer
+    def _cleanup_all_timers(self):
+        timers = self._timers
+        new_timers = self._new_timers
 
-    def add_timer(self, delay, callback):
-        return self.add_timer_absolute(delay + time.time(), callback)
+        while timers:
+            _, timer = timers.pop()
+            timer.timer_ended_cb()
+
+        # Although it is not the case with the current code base,
+        # the timers ended above may add new timers. So, the order
+        # is important.
+        while new_timers:
+            _, timer = new_timers.popleft()
+            timer.timer_ended_cb()
+
+    def check_loop(self):
+        raise NotImplementedError("check_loop")
+
+    def run_loop(self):
+        raise NotImplementedError("run_loop")
+
+    def wake_loop(self):
+        raise NotImplementedError("wake_loop")
+
+    def shutdown(self):
+        raise NotImplementedError("shutdown")
+
+
+class _WakeableLoop(_AbstractLoop):
+    _waker_class = _PipedWaker if os.name != 'nt' else _SocketedWaker
+
+    def __init__(self, map):
+        _AbstractLoop.__init__(self, map)
+        self.waker = self._waker_class(map)
+
+    def check_loop(self):
+        assert not self.waker.awake
+        self.wake_loop()
+        assert self.waker.awake
+        self.run_loop()
+        assert not self.waker.awake
+
+    def run_loop(self):
+        asyncore.loop(timeout=0.1, use_poll=True, map=self._map, count=1)
+
+    def wake_loop(self):
+        self.waker.wake()
```

Comment:
Does it make sense to check that the caller thread is different from the reactor thread? It doesn't make any sense to wake `poll` when we're already on the reactor thread.

Reply:
If we are in the reactor thread and calling this, we are already awake(not waiting on the poll call), so we are just writing a single byte to a pipe(or socket), which is almost no-op

Reply:
I don't think that it's not almost no-op. In case when there are no new external (unrelated with the waker pipe) events, writing into the waker will mean that the next `poll` will immediately exit (and we'll have to do two system calls instead of one), while it should lay dormant. I can see that Tornado has this check: https://github.com/tornadoweb/tornado/blob/branch5.1/tornado/ioloop.py#L1128

But this is definitely a nit, so it's up to you to implement the check or not.

Reply:
Added that